### DISCUSSION TITLE
Embeddeding the ultrasound messages in another audio file.

### DISCUSSION THREAD

**Discussion by quantaoflight** [ggerganov/ggwave#37] [Ideas]
Hi Greg, 
The ggwave library is quite exciting and I have tested it in various environments. Now I am trying to add the .wav file containing the message into any regular song which is in mp3 format. I am trying to use FFMPEG tool to do this, but the detection does not happen when the final output file is played out.

Any tool or software do you recommend to do this?

Thanks
MIhir

**Comment by ggerganov** [comment]
I haven't tried doing this, but my guess would be that somehow the mp3 compression is destroying the ultrasound signal.
I will try to make some experiments with this and see if there is a viable way to embed the ultrasound message in an mp3 file.

**Reply by ggerganov** [reply]
@quantaoflight 

So I tried various ways to encode the ggwave ultrasound messages into `mp3` format, but the transmission always fails. Looking at the spectrogram, indeed the compression ruins the signal and the receiver fails to decode the message.

However, I was able to encode an ultrasound message into `mp4` format.
Here is how to do it with ffmpeg:

```bash
# generate hello.wav using an ultrasound protocol
curl -sS 'https://ggwave-to-file.ggerganov.com/?m=Hello world!&p=4' --output hello.wav

# convert the generated .wav file to .mp4 file using ffmpeg
ffmpeg -i hello.wav -b:a 320k -q:a 0 -ar 48000 hello.mp4
```

Using this method I am able to reliably transmit an ultrasound message from my computer speakers by playing the generated `mp4` file with `vlc` and receive the message on my iPhone running the [waver](https://github.com/ggerganov/ggwave/tree/master/examples/waver) application.
